### PR TITLE
Expressions v1

### PR DESCRIPTION
**Pull request by daeMOn63**
This replaces the datalog Constraints with Expressions, aiming to provide more flexibility in the rules definition. Instead of being limited to predefined constraints, one can now compose custom constraints out of the available operations.
The old v0 constraints are compatible and are being converted to expressions on deserialization.

Expressions are evaluated when applying the rules (from biscuit verification or a query). They are composed of `Values` which are simply pushed in a simple stack, or Operations, which pop their arguments from the stack, process them and push their result onto the stack.
To be valid, expressions must leave a single boolean value at the end of their evaluation on the stack. When true, the rule then applies as before and generate its facts.

Some other related side changes:
- dropped the ability to generate v0 protobuf entities, allowing only v1 (because we can't really convert expressions back to constraint)
- the parser is dead. It will soon be updated in a separate PR, I'm waiting for the new syntax to stabilize a bit before updating.

### CODE REVIEW

File: converters_v1.go [hunk 1]
```diff
@@ -105,9 +103,9 @@ func tokenIDToProtoIDV1(input datalog.ID) (*pb.IDV1, error) {
 		expectedEltType := datalogSet[0].Type()
 		switch expectedEltType {
 		case datalog.IDTypeVariable:
-			return nil, errors.New("biscuit: failed to convert token ID to proto ID: set cannot contain variable")
+			return nil, errors.New("biscuit: failed to convert token ID to proto ID: set cannot contains variable")
```
Comment (titanous) [inline converters_v1.go:106]:
```suggestion
			return nil, errors.New("biscuit: failed to convert token ID to proto ID: set cannot contain variable")
```

File: converters_v1.go [hunk 2]
```diff
@@ -105,9 +103,9 @@ func tokenIDToProtoIDV1(input datalog.ID) (*pb.IDV1, error) {
...
 		case datalog.IDTypeSet:
-			return nil, errors.New("biscuit: failed to convert token ID to proto ID: set cannot contain other sets")
+			return nil, errors.New("biscuit: failed to convert token ID to proto ID: set cannot contains other sets")
```
Comment (titanous) [inline converters_v1.go:108]:
```suggestion
			return nil, errors.New("biscuit: failed to convert token ID to proto ID: set cannot contain other sets")
```

File: converters_v1.go [hunk 3]
```diff
@@ -166,9 +164,9 @@ func protoIDToTokenIDV1(input *pb.IDV1) (*datalog.ID, error) {
 		expectedEltType := reflect.TypeOf(elts[0].GetContent())
 		switch expectedEltType {
 		case reflect.TypeOf(&pb.IDV1_Variable{}):
-			return nil, errors.New("biscuit: failed to convert proto ID to token ID: set cannot contain variable")
+			return nil, errors.New("biscuit: failed to convert proto ID to token ID: set cannot contains variable")
```
Comment (titanous) [inline converters_v1.go:167]:
```suggestion
			return nil, errors.New("biscuit: failed to convert proto ID to token ID: set cannot contain variable")
```

File: converters_v1.go [hunk 4]
```diff
@@ -166,9 +164,9 @@ func protoIDToTokenIDV1(input *pb.IDV1) (*datalog.ID, error) {
...
 		case reflect.TypeOf(&pb.IDV1_Set{}):
-			return nil, errors.New("biscuit: failed to convert proto ID to token ID: set cannot contain other sets")
+			return nil, errors.New("biscuit: failed to convert proto ID to token ID: set cannot contains other sets")
```
Comment (titanous) [inline converters_v1.go:169]:
```suggestion
			return nil, errors.New("biscuit: failed to convert proto ID to token ID: set cannot contain other sets")
```

File: datalog/expressions.go [hunk 5]
```diff
@@ -0,0 +1,674 @@
...
+func (e *Expression) Evaluate(values map[Variable]*ID) (ID, error) {
+	s := &stack{}
+
+	for _, op := range *e {
+		switch op.Type() {
+		case OpTypeValue:
+			id := op.(Value).ID
+			switch id.Type() {
+			case IDTypeVariable:
+				var ok bool
```
Comment (titanous) [inline datalog/expressions.go:31]: This variable definition looks redundant?

File: datalog/expressions.go [hunk 6]
```diff
@@ -0,0 +1,674 @@
...
+
+			res, err := op.(UnaryOp).Eval(v)
+			if err != nil {
+				return nil, fmt.Errorf("datalog: expressions: unary eval failed: %w", err)
+			}
+			s.Push(res)
+		case OpTypeBinary:
+			right, err := s.Pop()
+			if err != nil {
+				return nil, fmt.Errorf("datalog: expressiosn: failed to pop binary right value: %w", err)
```
Comment (titanous) [inline datalog/expressions.go:53]:
```suggestion
				return nil, fmt.Errorf("datalog: expressions: failed to pop binary right value: %w", err)
```

File: datalog/expressions.go [hunk 7]
```diff
@@ -0,0 +1,674 @@
...
+			}
+			left, err := s.Pop()
+			if err != nil {
+				return nil, fmt.Errorf("datalog: expressiosn: failed to pop binary left value: %w", err)
```
Comment (titanous) [inline datalog/expressions.go:57]:
```suggestion
				return nil, fmt.Errorf("datalog: expressions: failed to pop binary left value: %w", err)
```

File: datalog/expressions.go [hunk 8]
```diff
@@ -0,0 +1,674 @@
...
+}
+
+type UnaryOpType byte
+
+const (
+	UnaryNegate UnaryOpType = iota
+	UnaryParens
+)
+
+// Negate returns the negation of value
```
Comment (titanous) [inline datalog/expressions.go:179]:
```suggestion
// Negate returns the negation of a value.
```

File: datalog/expressions.go [hunk 9]
```diff
@@ -0,0 +1,674 @@
...
+		out = !value.(Bool)
+	default:
+		return nil, fmt.Errorf("datalog: unexpected Negate value type: %d", value.Type())
+	}
+
+	return out, nil
+}
+
+// Parens allows expression priority and grouping (like parenthesis in math operations)
+// its a no op, but is used to print back the expressions properly, putting their value
```
Comment (titanous) [inline datalog/expressions.go:199]:
```suggestion
// it is a no-op, but is used to print back the expressions properly, putting their value
```

File: datalog/expressions.go [hunk 10]
```diff
@@ -0,0 +1,674 @@
...
+	BinaryAdd
+	BinarySub
+	BinaryMul
+	BinaryDiv
+	BinaryAnd
+	BinaryOr
+)
+
+// LessThan returns true when left is less than right.
+// It requires left and right to have the same concret type
```
Comment (titanous) [inline datalog/expressions.go:282]:
```suggestion
// It requires left and right to have the same concrete type
```

File: datalog/expressions.go [hunk 11]
```diff
@@ -0,0 +1,674 @@
...
+// and only accepts Integer
```
Comment (titanous) [inline datalog/expressions.go:283]:
```suggestion
// and only accepts Integer.
```

File: datalog/expressions.go [hunk 12]
```diff
@@ -0,0 +1,674 @@
...
+	default:
+		return nil, fmt.Errorf("datalog: unexpected LessThan value type: %d", left.Type())
+	}
+
+	return out, nil
+}
+
+// LessOrEqual returns true when left is less or equal right.
+// It requires left and right to have the same concret type
+// and only accepts Integer and Date
```
Comment (titanous) [inline datalog/expressions.go:307]:
```suggestion
// and only accepts Integer and Date.
```

File: datalog/expressions.go [hunk 13]
```diff
@@ -0,0 +1,674 @@
...
+		out = Bool(left.(Integer) < right.(Integer))
+	default:
+		return nil, fmt.Errorf("datalog: unexpected LessThan value type: %d", left.Type())
+	}
+
+	return out, nil
+}
+
+// LessOrEqual returns true when left is less or equal right.
+// It requires left and right to have the same concret type
```
Comment (titanous) [inline datalog/expressions.go:306]:
```suggestion
// It requires left and right to have the same concrete type
```

File: datalog/expressions.go [hunk 14]
```diff
@@ -0,0 +1,674 @@
...
+	case IDTypeInteger:
+		out = Bool(left.(Integer) < right.(Integer))
+	default:
+		return nil, fmt.Errorf("datalog: unexpected LessThan value type: %d", left.Type())
+	}
+
+	return out, nil
+}
+
+// LessOrEqual returns true when left is less or equal right.
```
Comment (titanous) [inline datalog/expressions.go:305]:
```suggestion
// LessOrEqual returns true when left is less or equal than right.
```

File: datalog/expressions.go [hunk 15]
```diff
@@ -0,0 +1,674 @@
...
+		out = Bool(left.(Date) <= right.(Date))
+	default:
+		return nil, fmt.Errorf("datalog: unexpected LessOrEqual value type: %d", left.Type())
+	}
+
+	return out, nil
+}
+
+// GreaterThan returns true when left is greater than right.
+// It requires left and right to have the same concret type
```
Comment (titanous) [inline datalog/expressions.go:332]:
```suggestion
// It requires left and right to have the same concrete type
```

File: datalog/expressions.go [hunk 16]
```diff
@@ -0,0 +1,674 @@
...
+// and only accepts Integer
```
Comment (titanous) [inline datalog/expressions.go:333]:
```suggestion
// and only accepts Integer.
```

File: datalog/expressions.go [hunk 17]
```diff
@@ -0,0 +1,674 @@
...
+		out = Bool(left.(Integer) > right.(Integer))
+	default:
+		return nil, fmt.Errorf("datalog: unexpected GreaterThan value type: %d", left.Type())
+	}
+
+	return out, nil
+}
+
+// GreaterOrEqual returns true when left is greater than right.
+// It requires left and right to have the same concret type
```
Comment (titanous) [inline datalog/expressions.go:356]:
```suggestion
// It requires left and right to have the same concrete type
```

File: datalog/expressions.go [hunk 18]
```diff
@@ -0,0 +1,674 @@
...
+// and only accepts Integer and Date
```
Comment (titanous) [inline datalog/expressions.go:357]:
```suggestion
// and only accepts Integer and Date.
```

File: datalog/expressions.go [hunk 19]
```diff
@@ -0,0 +1,674 @@
...
+		out = Bool(left.(Date) >= right.(Date))
+	default:
+		return nil, fmt.Errorf("datalog: unexpected GreaterOrEqual value type: %d", left.Type())
+	}
+
+	return out, nil
+}
+
+// Equal returns true when left and right are equal.
+// It requires left and right to have the same concret type
```
Comment (titanous) [inline datalog/expressions.go:382]:
```suggestion
// It requires left and right to have the same concrete type
```

File: datalog/expressions.go [hunk 20]
```diff
@@ -0,0 +1,674 @@
...
+// and only accepts Integer, Bytes or String
```
Comment (titanous) [inline datalog/expressions.go:383]:
```suggestion
// and only accepts Integer, Bytes or String.
```

File: datalog/expressions.go [hunk 21]
```diff
@@ -0,0 +1,674 @@
...
+	case IDTypeString:
+	default:
+		return nil, fmt.Errorf("datalog: unexpected Equal value type: %d", left.Type())
+	}
+
+	return Bool(left.Equal(right)), nil
+}
+
+// Contains returns true when right value exists in the left Set.
+// right value must be an Integer, Bytes, String or Symbol
```
Comment (titanous) [inline datalog/expressions.go:406]:
```suggestion
// The right value must be an Integer, Bytes, String or Symbol.
```

File: datalog/expressions.go [hunk 22]
```diff
@@ -0,0 +1,674 @@
...
+// left value must be a Set, containing elements of right type.
```
Comment (titanous) [inline datalog/expressions.go:407]:
```suggestion
// The left value must be a Set, containing elements of right type.
```

File: datalog/expressions.go [hunk 23]
```diff
@@ -0,0 +1,674 @@
...
+	case IDTypeBytes:
+	case IDTypeString:
+	default:
+		return nil, fmt.Errorf("datalog: unexpected Equal value type: %d", left.Type())
+	}
+
+	return Bool(left.Equal(right)), nil
+}
+
+// Contains returns true when right value exists in the left Set.
```
Comment (titanous) [inline datalog/expressions.go:405]:
```suggestion
// Contains returns true when the right value exists in the left Set.
```

File: datalog/expressions.go [hunk 24]
```diff
@@ -0,0 +1,674 @@
...
+		}
+		if right.Equal(elt) {
+			return Bool(true), nil
+		}
+	}
+
+	return Bool(false), nil
+}
+
+// Prefix returns true when left string starts with right string
```
Comment (titanous) [inline datalog/expressions.go:440]:
```suggestion
// Prefix returns true when the left string starts with the right string.
```

File: datalog/expressions.go [hunk 25]
```diff
@@ -0,0 +1,674 @@
...
+// left and right must be String
```
Comment (titanous) [inline datalog/expressions.go:441]:
```suggestion
// left and right must be String.
```

File: datalog/expressions.go [hunk 26]
```diff
@@ -0,0 +1,674 @@
...
+	}
+	sright, ok := right.(String)
+	if !ok {
+		return nil, fmt.Errorf("datalog: Prefix requires right value to be a String, got %T", right)
+	}
+
+	return Bool(strings.HasPrefix(string(sleft), string(sright))), nil
+}
+
+// Suffix returns true when left string ends with right string
```
Comment (titanous) [inline datalog/expressions.go:460]:
```suggestion
// Suffix returns true when the left string ends with the right string.
```

File: datalog/expressions.go [hunk 27]
```diff
@@ -0,0 +1,674 @@
...
+// left and right must be String
```
Comment (titanous) [inline datalog/expressions.go:461]:
```suggestion
// left and right must be String.
```

File: datalog/expressions.go [hunk 28]
```diff
@@ -0,0 +1,674 @@
...
+	}
+	sright, ok := right.(String)
+	if !ok {
+		return nil, fmt.Errorf("datalog: Suffix requires right value to be a String, got %T", right)
+	}
+
+	return Bool(strings.HasSuffix(string(sleft), string(sright))), nil
+}
+
+// Regex returns true when right string is a regexp and left match against it.
```
Comment (titanous) [inline datalog/expressions.go:480]:
```suggestion
// Regex returns true when the right string is a regexp and left matches against it.
```

File: datalog/expressions.go [hunk 29]
```diff
@@ -0,0 +1,674 @@
...
+// left and right must be String
```
Comment (titanous) [inline datalog/expressions.go:481]:
```suggestion
// left and right must be String.
```

File: datalog/expressions.go [hunk 30]
```diff
@@ -0,0 +1,674 @@
...
+	}
+
+	re, err := regexp.Compile(string(sright))
+	if err != nil {
+		return nil, fmt.Errorf("datalog: invalid regex: %q: %v", right, err)
+	}
+	return Bool(re.Match([]byte(sleft))), nil
+}
+
+// Add performs the addition of left + right and returns the result
```
Comment (titanous) [inline datalog/expressions.go:504]:
```suggestion
// Add performs the addition of left + right and returns the result.
```

File: datalog/expressions.go [hunk 31]
```diff
@@ -0,0 +1,674 @@
...
+// It requires left and right to be Integer
```
Comment (titanous) [inline datalog/expressions.go:505]:
```suggestion
// It requires left and right to be Integer.
```

File: datalog/expressions.go [hunk 32]
```diff
@@ -0,0 +1,674 @@
...
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+	"regexp"
+	"strings"
+)
+
+// maxStackSize defines the maximum number of elements that can be  stored on the stack.
```
Comment (titanous) [inline datalog/expressions.go:11]:
```suggestion
// maxStackSize defines the maximum number of elements that can be stored on the stack.
```

File: datalog/expressions.go [hunk 33]
```diff
@@ -0,0 +1,674 @@
...
+// Trying to store more than maxStackSize elements returns an error
```
Comment (titanous) [inline datalog/expressions.go:12]:
```suggestion
// Trying to store more than maxStackSize elements returns an error.
```

File: datalog/expressions.go [hunk 34]
```diff
@@ -0,0 +1,674 @@
...
+	res := big.NewInt(0)
+	res.Add(bleft, bright)
+
+	if !res.IsInt64() {
+		return nil, ErrInt64Overflow
+	}
+	return Integer(res.Int64()), nil
+}
+
+// Sub performs the substraction of left - right and returns the result
```
Comment (titanous) [inline datalog/expressions.go:532]:
```suggestion
// Sub performs the substraction of left - right and returns the result.
```

File: datalog/expressions.go [hunk 35]
```diff
@@ -0,0 +1,674 @@
...
+// It requires left and right to be Integer
```
Comment (titanous) [inline datalog/expressions.go:533]:
```suggestion
// It requires left and right to be Integer.
```

File: datalog/expressions.go [hunk 36]
```diff
@@ -0,0 +1,674 @@
...
+	res := big.NewInt(0)
+	res.Sub(bleft, bright)
+
+	if !res.IsInt64() {
+		return nil, ErrInt64Overflow
+	}
+	return Integer(res.Int64()), nil
+}
+
+// Mul performs the multiplication of left * right and returns the result
```
Comment (titanous) [inline datalog/expressions.go:560]:
```suggestion
// Mul performs the multiplication of left * right and returns the result.
```

File: datalog/expressions.go [hunk 37]
```diff
@@ -0,0 +1,674 @@
...
+// It requires left and right to be Integer
```
Comment (titanous) [inline datalog/expressions.go:561]:
```suggestion
// It requires left and right to be Integer.
```

File: datalog/expressions.go [hunk 38]
```diff
@@ -0,0 +1,674 @@
...
+	res.Mul(bleft, bright)
+
+	if !res.IsInt64() {
+		return nil, ErrInt64Overflow
+	}
+
+	return Integer(res.Int64()), nil
+}
+
+// Div performs the division of left / right and returns the result
```
Comment (titanous) [inline datalog/expressions.go:589]:
```suggestion
// Div performs the division of left / right and returns the result.
```

File: datalog/expressions.go [hunk 39]
```diff
@@ -0,0 +1,674 @@
...
+// It requires left and right to be Integer
```
Comment (titanous) [inline datalog/expressions.go:590]:
```suggestion
// It requires left and right to be Integer.
```

File: datalog/expressions.go [hunk 40]
```diff
@@ -0,0 +1,674 @@
...
+	}
+
+	if iright == 0 {
+		return nil, ErrExprDivByZero
+	}
+
+	return Integer(ileft / iright), nil
+}
+
+// And performs a logical AND between left and right and returns a Bool
```
Comment (titanous) [inline datalog/expressions.go:613]:
```suggestion
// And performs a logical AND between left and right and returns a Bool.
```

File: datalog/expressions.go [hunk 41]
```diff
@@ -0,0 +1,674 @@
...
+// It requires left and right to be Bool
```
Comment (titanous) [inline datalog/expressions.go:614]:
```suggestion
// It requires left and right to be Bool.
```

File: datalog/expressions.go [hunk 42]
```diff
@@ -0,0 +1,674 @@
...
+	}
+	bright, ok := right.(Bool)
+	if !ok {
+		return nil, fmt.Errorf("datalog: And requires right value to be a Bool, got %T", right)
+	}
+
+	return Bool(bleft && bright), nil
+}
+
+// Or performs a logical OR between left and right and returns a Bool
```
Comment (titanous) [inline datalog/expressions.go:633]:
```suggestion
// Or performs a logical OR between left and right and returns a Bool.
```

File: datalog/expressions.go [hunk 43]
```diff
@@ -0,0 +1,674 @@
...
+// It requires left and right to be Bool
```
Comment (titanous) [inline datalog/expressions.go:634]:
```suggestion
// It requires left and right to be Bool.
```

File: parser/parser_test.go [hunk 44]
```diff
@@ -57,434 +53,434 @@ func getFactTestCases() []testCase {
 	}
 }
 
-func getRuleTestCases() []testCase {
-	t1, _ := time.Parse(time.RFC3339, time.Now().Format(time.RFC3339))
-	t2, _ := time.Parse(time.RFC3339, time.Now().Add(2*time.Second).Format(time.RFC3339))
+// func getRuleTestCases() []testCase {
```
Comment (titanous) [inline parser/parser_test.go:56]: I assume these tests need to be rewritten?

Comment (daeMOn63) [reply parser/parser_test.go:56]: Yes, the parser is totally broken right now, and there are still some ongoing changes to the syntax of the new expressions. I'll fix it when we'll have a more stable syntax.

**Comment by daeMOn63** [comment]
@titanous do you still have some pending reviews on this one?

**Review by titanous — APPROVED** [review]

**Pull request merged**